src/server.h

```diff
@@ -429,6 +431,23 @@ extern int configOOMScoreAdjValuesDefaults[CONFIG_OOM_COUNT];
 #define CLIENT_REPLICATION_DONE (1ULL << 51)         /* Indicate that replication has been done on the client */
 #define CLIENT_AUTHENTICATED (1ULL << 52)            /* Indicate a client has successfully authenticated */
 
+#define CLIENT_REPL_MAIN_CONN (1ULL<<52) /* RDB connection: track a connection
+                                                 which is used for online replication data */
+#define CLIENT_REPL_RDB_CONN (1ULL<<53) /* RDB connection: track a connection
+                                                 which is used for rdb snapshot */
+#define CLIENT_PROTECTED_RDB_CONN (1ULL<<54) /* RDB connection: Protects the RDB client from premature 
+                    * release during full sync. This flag is used to ensure that the RDB client, which 
+                    * references the first replication data block required by the replica, is not 
+                    * released prematurely. Protecting the client is crucial for prevention of 
+                    * synchronization failures:
+                    * If the RDB client is released before the replica initiates PSYNC, the primary 
+                    * will reduce the reference count (o->refcount) of the block needed by the replica. 
+                    * This could potentially lead to the removal of the required data block, resulting 
+                    * in synchronization failures. Such failures could occur even in scenarios where
+                    * the replica only needs an additional 4KB beyond the minimum size of the repl_backlog.
+                    * By using this flag, we ensure that the RDB client remains intact until the replica 
+                    * has successfully initiated PSYNC. */
```

Comment:
>PSYNC can only be initiated after the background process has started generating the RDB, as this is the first point we know the end offset of the RDB. 

Here is my understanding of the sequence:

1. Parent process sends $ENDOFF https://github.com/valkey-io/valkey/pull/60/files#diff-c77a3d2b15213159471dad3359f23629c2297c3579861945e94ff05c34bb3d7dR3558
2. Parent process forks https://github.com/valkey-io/valkey/pull/60/files#diff-c77a3d2b15213159471dad3359f23629c2297c3579861945e94ff05c34bb3d7dR3569
3. Replica receives $ENDOFF https://github.com/valkey-io/valkey/pull/60/files#diff-955c0fee32b574a1d9adb4540b8ff1564b15db9695d88a8dabc66b3c19c825daR2645
4. Replica asynchronously sends PSYNC https://github.com/valkey-io/valkey/pull/60/files#diff-955c0fee32b574a1d9adb4540b8ff1564b15db9695d88a8dabc66b3c19c825daR2677

I think we know the end offset before forking the process.

> The RDB generation process needs to begin asynchronously to avoid blocking the main thread.

I assume you meant the **replica** main thread? I think this would be fine since 1) replica isn't online; 2) the block should be short.